### PR TITLE
Add a pending remediation status and support tracking opened PRs

### PR DESCRIPTION
**Pull request by rdimitrov**
# Summary

The following PR:

- Adds a new metadata column in the `rule_details_remediate` table for storing the opened PR number
- Refactors the way we process remediations through PRs - we no longer use magic comments to keep track of opened ones.
- If the issue is fixed out of band (through another PR for example) we'll now automatically close the PR along with the opened alert too
- Status is now `pending` when evaluation failed and remediation opened a PR
- Status is `skipped` if we don't need remediation
- Status is `success` if remediation was considered instant (i.e. `rest`) but we haven't re-evaluated yet (so rule is still failing). 
- Once we re-evaluate and the issue is fixed, the remediation status will become `skipped`.
- Adds a new `ClosePullRequest()` handler
- Removes the part where we were not updating the database for alerts with skipped status
- Refactors the part where we make the decision of whether we should remediate or not to happen in one place in the actions engine
- Fixes another bug where the remediation for pinned actions actually opens another PR after merging the first one

**What's left (can be done in a follow-up PR):**
- The PR adds more log messages, but I will revisit/update their usefulness 
- Currently if there's nothing to do, we return the previous action state as we do upsert at the end. The good thing is it updates the timestamp, but not sure how wasteful it is. Perhaps it's better to return a custom ErrActionDidNothing and skip upserting if this is the case. 
- Test this more with other ruletypes that remediate through a PR
- Add the pull request URL to be returned by the API
- Support rebasing/updating the PR if main changes while the PR is opened

Supersedes https://github.com/stacklok/minder/pull/2530
Fixes https://github.com/stacklok/minder/issues/2526
Fixes https://github.com/stacklok/minder/issues/1207
Fixes https://github.com/stacklok/minder/issues/1201

## Change Type

***Mark the type of change your PR introduces:***

- [x] Bug fix (resolves an issue without affecting existing features)
- [x] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [x] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

Tested locally with a demo repository using the `dependabot_configured` ruletype:

**Start with a failing repo:**
* evaluation is failure
* rem is pending
* alert is on
* We have an opened PR and alert

**Case 1 - Merged the PR**
* eval is success
* rem is skipped
* alert is off
* alert is closed automatically

**Case 2 - Fixed the issue through another PR**
* Eval is ok
* rem is skipped
* alert is off
* both PR and alert are closed automatically

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [x] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: internal/providers/github/common.go
Comment (jhrozek) [inline internal/providers/github/common.go:572]: Instead of hand-crafting the close you can use PullRequest.Edit and set the state to closed I think

File: database/query/profile_status.sql
Comment (jhrozek) [inline database/query/profile_status.sql:57]: I think I understand what you are doing now but I can't read the previous code anymore :-) why did we code it like this?

File: internal/engine/actions/actions.go
Comment (jhrozek) [inline internal/engine/actions/actions.go:188]: aren't there legitimate cases for the PR evaluator where we want to push a new version of the PR even if the evaluation was failure before and is still failing?
e.g. I want frizbee to unroll tags to shas. The evaluation is failing and there is a PR. While the PR is open, someone pushes another workflow with more actions..

File: internal/engine/actions/actions.go
Comment (jhrozek) [inline internal/engine/actions/actions.go:211]: can we even get pending eval here (its fine to leave it there, just asking)

File: database/query/profile_status.sql
Comment (evankanderson) [reply database/query/profile_status.sql:57]: Previous comments assume a bunch more context than you might have now:

https://github.com/stacklok/minder/pull/1228/files#r1365307972

File: internal/engine/actions/actions.go
Comment (evankanderson) [inline internal/engine/actions/actions.go:145]: This comment doesn't quite line up with the code below, and the code below seems clear without it.  Remove the comment?

File: internal/engine/actions/actions.go
Comment (evankanderson) [inline internal/engine/actions/actions.go:157]: Do we need a default case?

Comment (evankanderson) [inline internal/engine/actions/actions.go:157]: The above might read more easily if you extract the `params.GetEvalStatusFromDb()` that's used 4 times into an 'evalStatus' temporary.

```suggestion
		logger.Msg("skipping action")
		evalStatus := params.GetEvalStatusFromDb()
		switch actionType {
		case remediate.ActionType:
			// Return the previous remediation status.
			err := enginerr.RemediationStatusAsError(evalStatus.RemStatus.RemediationStatusTypes)
			return *(getMeta(evalStatus.RemMetadata)), err
		case alert.ActionType:
			// Return the previous alert status.
			err := enginerr.AlertStatusAsError(evalStatus.AlertStatus.AlertStatusTypes)
			return *(getMeta(evalStatus.AlertMetadata)), err
		}
```

Comment (evankanderson) [reply internal/engine/actions/actions.go:157]: In particular, if there's a different action type, it drops out of the `if` and goes on to the rest of the body.

File: internal/engine/actions/actions.go
Comment (evankanderson) [inline internal/engine/actions/actions.go:283]: Is this logic the same as 191-216?

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:188]: That's a good point, I haven't thought about the frizbee action as I assumed we are proposing only new files so far (dependabot, codeql, trivy, etc). I'll try to think about how to implement that 👍 Do you know if it's working like that right now and this will break it or?

File: internal/providers/github/common.go
Comment (rdimitrov) [reply internal/providers/github/common.go:572]: ah, missed that, thanks! 👍

File: database/query/profile_status.sql
Comment (rdimitrov) [reply database/query/profile_status.sql:57]: yeah, I think it was probably wrong to do it like that in the first place. Now it should act the same for remediations and alerts and the logs should reflect the exact state we processed 👍

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:211]: I don't think so too, but the lint check said it's a good practice so thus why I added it 👍

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:145]: good point, I'll remove it 👍

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:157]: My assumption was every action should be able to act correctly when it's called with the `ActionCmdDoNothing ` command, so even if we add a new one later and we miss to add it to this part there shouldn't be a problem. But now that you asked it it got me thinking that if we end up going to the action while skipping this part for some reason currently this will not preserve the old state.

I'll remove the `switch` here and add each case to be handled explicitly by each action. This way it's up to the action to implement what it should do on `ActionCmdDoNothing`, i.e. some may decide to preserve old state, some may choose something else.

Thanks!

File: internal/engine/actions/actions.go
Comment (jhrozek) [reply internal/engine/actions/actions.go:188]: with frizbee in particular I think the additional changes are the way it works now, yes, because unllike dependabot and codeql where we just put in the content, with frizbee you feed the content you get from evaluation to the frizbee library which modifies it and gives you back the modified content.

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:283]: Yes, I decided to refactor the separate `if` statements into a single switch now that we have duplicate behaviour for different evaluation statuses, i.e. `db.EvalStatusTypesError` and `db.EvalStatusTypesFailure`

**Comment by rdimitrov** [comment]
These are the rest of the changes I think we can do in a follow-up PR:
* The PR adds more log messages, but I will revisit/update their usefulness
* Add the pull request URL to be returned by the API
* Support rebasing/updating the remediation PR if main changes while the PR is opened

File: database/migrations/000040_remediate_metadata.down.sql
Comment (jhrozek) [inline database/migrations/000040_remediate_metadata.down.sql:1]: please rename the migrations to 41

File: internal/engine/actions/actions.go
Comment (jhrozek) [inline internal/engine/actions/actions.go:118]: since we always pass metadata for both remediate and alerts, would it make sense to change the interface for `actions.Do` to accept a `json.RawMessage`, not a pointer to it to avoid crashes in some corner cases?

File: internal/engine/actions/actions.go
Comment (jhrozek) [inline internal/engine/actions/actions.go:161]: Should this say `RemStatus.Valid`?

File: database/migrations/000040_remediate_metadata.down.sql
Comment (rdimitrov) [reply database/migrations/000040_remediate_metadata.down.sql]: Done 👍

File: internal/engine/actions/actions.go
Comment (jhrozek) [reply internal/engine/actions/actions.go:118]: (not in this PR)

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go]: Good catch! 💯

File: internal/engine/actions/actions.go
Comment (rdimitrov) [reply internal/engine/actions/actions.go:118]: good point, we certainly can 👍

File: internal/engine/actions/actions.go
Comment (jhrozek) [reply internal/engine/actions/actions.go:170]: I'm fine fixing this in a follow-up, but could this be solved by:
 - if the old status is failure always defer to the remediator
 - in the PR remediator, check what HEAD the PR points to. If it is the same as the hash of the commit we created locally, shortcut, otherwise go forward and force-push
 - the other remediators we have at least for now are "atomic" so this should not happen in the first place and if it does, then we should have anotoher mechanism that prevents an infinite remediate loop anyway

**Review by jhrozek — APPROVED** [review]
Thank you the logic reads much better now. I'm fine merging this and following up on this work with figuring out how to update the pending remediations.

**Comment by rdimitrov** [comment]
Thanks! 

For anyone else looking at the PR, let's wait a bit until we merge this so we have the smoke test updates in a PR ready for merging too 👍

**Pull request merged**
